### PR TITLE
Add buffer to speed up transaction streaming

### PR DESCRIPTION
**Pull request by rautenrieth-da**
This PR fixes an issue that was causing the transaction service to become almost twice as slow. I still don't understand what exactly is going on, but this small PR restores the original performance. 

## Changelog

```
CHANGELOG_BEGIN
- [Ledger API Server] Fixed an issue that was causing a slowdown of the
  transaction service.
CHANGELOG_END
```

## Measuring the impact

Method:
- Generate an index database with huntsman (many thanks to Martino)
- Build ledger-on-memory from commit 9ed787cb
- Run ledger-on-memory in ledger-api-only mode, connect to the existing index database
- Run spider-mock to measure the speed of the transaction service
- Repeat with ledger-on-memory where the out-of-order fix was reverted

## Results

Reading 17009 transactions took:
```
mergeSubstreams():                 18.602 secs (before the out-of-order fix)
concatSubstreams():                29.007 secs (after the out-of-order fix)
statefulMapConcat():               28.656 secs (first attempt: code without substreams)
statefulMapConcat() + 2x buffer(): 18.767 secs (second attempt: add buffers)
concatSubstreams() + buffer():     18.981 secs (third attempt: smallest change)
```

Code for the `statefulMapConcat` versions: 
```
source
  .map(Some(_))
  .concat(Source.single(None))
  .statefulMapConcat(() => {
    var previousSegmentKey: K = null.asInstanceOf[K]
    val builder = new VectorBuilder[A]
    entryO => {
      entryO match {
        case Some(entry) =>
          val keyForEntry = by(entry)
          if (keyForEntry != previousSegmentKey) {
            previousSegmentKey = keyForEntry
            val result = builder.result()
            builder.clear()
            builder += entry
            List(result)
          } else {
            builder += entry
            List.empty
          }
        case None =>
          List(builder.result())
      }
    }
  })
```

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/package.scala [hunk 1]
```diff
@@ -74,6 +75,7 @@ package object events {
       .map(_._1)
       .fold(Vector.empty[A])(_ :+ _)
       .concatSubstreams
+      .buffer(128, OverflowStrategy.backpressure)
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/package.scala:78]: `.buffer(16)` did not bring much improvement. I did not investigate other buffer sizes.

Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/store/dao/events/package.scala:78]: Can you pull this out into a constant somewhere? Or better yet, a parameter?

I worry about the volume of magic numbers inside our codebase. I'd really like to push them to at the very least, a single file with all the magic numbers.

**Review by stefanobaghino-da — APPROVED** [review]
What @SamirTalwar said, plus I would recommend also adding the reason for the buffer. There already is an "implementation detail" section.

Thank you so much for the very accurate analysis.

**Comment by SamirTalwar-DA** [comment]
> Thank you so much for the very accurate analysis.

I second this. It was very helpful!

**Comment by rautenrieth-da** [comment]
@stefanobaghino-da @SamirTalwar-DA I have moved the `.buffer()` call to TransactionReader and have created a documented private val for the magic number. I felt the buffer call does not belong to the generic `groupContiguous` method. I have verified that the performance of the transaction stream is slightly better than before the fix:

```
mergeSubstreams():             16.7 secs (before the out-of-order fix)
concatSubstreams():            26.6 secs (after the out-of-order fix)
concatSubstreams() + buffer(): 16.4 secs (after this PR)
```

**Comment by stefanobaghino-da** [comment]
> @stefanobaghino-da @SamirTalwar-DA I have moved the `.buffer()` call to TransactionReader and have created a documented private val for the magic number. I felt the buffer call does not belong to the generic `groupContiguous` method. I have verified that the performance of the transaction stream is slightly better than before the fix:
> 
> ```
> mergeSubstreams():             16.7 secs (before the out-of-order fix)
> concatSubstreams():            26.6 secs (after the out-of-order fix)
> concatSubstreams() + buffer(): 16.4 secs (after this PR)
> ```

Given the hard-coded number it probably makes more sense like this. Do you know what `.mergeSubstreams` does?

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsReader.scala [hunk 2]
```diff
@@ -50,6 +51,10 @@ private[dao] final class TransactionsReader(
 
   private val logger = ContextualizedLogger.get(this.getClass)
 
+  // TransactionReader adds an Akka stream buffer at the end of all streaming queries.
+  // This significantly improves the performance of the transaction service.
+  private val outputStreamBufferSize = 128
```
Comment (SamirTalwar-DA) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsReader.scala:56]: I would still prefer if we moved constants like this into a single file, but I'm OK with this for now.

**Review by SamirTalwar-DA — APPROVED** [review]

**Comment by rautenrieth-da** [comment]
> Do you know what .mergeSubstreams does?

I don't, unfortunately. All I know is that concatSubstreams() is implemented as mergeSubstreams() with maxParallelism=1.

**Pull request merged**